src/stages/agent.ts

```diff
@@ -57,47 +69,308 @@ export class Agent implements Stage<Agent.Input, Agent.Output> {
+            {
+              role: "user",
+              content: {
+                type: "text",
+                text: validationFailure.feedback,
+              },
+            },
+          ]
+        : []
+    ) satisfies Message[];
 
-    throw new Error("Method not implemented.");
+    return [
+      ...baseMessage,
+      ...prevMessages,
+      ...queryMessage,
+      ...validationPrompt,
+    ];
+  }
+
+  private onText(text: CompletionTextMessage): AgentAction {
+    return {
+      type: "chat",
+      message: text.text,
+    };
   }
-}
 
-// TODO: dummy interface
-async function callLlm(
-  _stage: unknown,
-  _ctx: StageContext,
-  _input: Agent.Input,
-  _prompts: Record<string, string>,
-  _request: Omit<LmBridgeRequest, "messages">
-): Promise<AgentAction[]> {
-  return [];
+  private async onToolUse(
+    ctx: StageContext,
+    toolUse: CompletionToolUseMessage
+  ): Promise<AgentAction> {
+    switch (toolUse.toolName) {
+      case "lookup_functions": {
+        const validatedArguments = validate<{
+          thoughts: string;
+          queries: {
+            query: string;
+            specifications?: string;
+          }[];
+        }>(toolUse.arguments);
+        if (!validatedArguments.success) {
+          throw new Error(
+            `[parsing JSON] your response contains invalid syntax:\n${validatedArguments.errors}`
+          );
+        }
+
+        const { thoughts, queries } = validatedArguments.data;
+        return {
+          type: "lookup_functions",
+          functionCall: toolUse,
+          thoughts,
+          queries: queries.map((v) => {
+            return {
+              query: v.query,
+              specifications: v.specifications,
+            };
+          }),
+        };
+      }
+      case "run_functions": {
+        const validatedArguments = validate<{
+          thoughts: string;
+          items: {
+            purpose: string;
+            functionId: string;
+          }[];
+        }>(toolUse.arguments);
+        if (!validatedArguments.success) {
+          throw new Error(
+            `[parsing JSON] your response contains invalid syntax:\n${validatedArguments.errors}`
+          );
+        }
+        const { thoughts, items } = validatedArguments.data;
+        const functions = await Promise.all(
+          items.map(async (v) => {
+            const func = await ctx.findFunction(ctx.sessionId, v.functionId);
+            if (!func) {
+              throw new Error(
+                `your response contains an invalid function id \`${v.functionId}\`; which does not exist in the list of available functions`
+              );
+            }
+            return {
+              purpose: v.purpose,
+              function: func,
+            };
+          })
+        );
+
+        return {
+          type: "run_functions",
+          functionCall: toolUse,
+          thoughts,
+          items: functions,
+        };
+      }
+
+      default:
+        throw new Error(
```

Comment:
Stage 안에서 발생하는 에러 중, 에이전트가 recover할 수 있는 종류의 에러는 `StageError`를 사용하면 좋습니다.

```suggestion
        throw new StageError(
```